include/ibamr/BrinkmanPenalizationRigidBodyDynamics.h

```diff
@@ -0,0 +1,372 @@
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef included_BrinkmanPenalizationRigidBodyDynamics
+#define included_BrinkmanPenalizationRigidBodyDynamics
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include "ibamr/BrinkmanPenalizationStrategy.h"
+
+#include "ibtk/ibtk_utilities.h"
+
+#include "tbox/Pointer.h"
+
+#include "Eigen/Core"
+#include "Eigen/Geometry"
+
+/////////////////////////////// NAMESPACE ////////////////////////////////////
+
+namespace IBAMR
+{
+class IBHydrodynamicSurfaceForceEvaluator;
+class INSVCStaggeredHierarchyIntegrator;
+class AdvDiffHierarchyIntegrator;
+} // namespace IBAMR
+namespace SAMRAI
+{
+namespace pdat
+{
+template <int DIM, class TYPE>
+class CellVariable;
+} // namespace pdat
+} // namespace SAMRAI
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBAMR
+{
+/*!
+ * \brief BrinkmanPenalizationRigidBodyDynamics provides an implementation of Brinkman penalization
+ * body force for a rigid body motion in the momentum equation.
+ *
+ * The penalization force is taken to be \f$ \frac{\chi}{\kappa}(\bm{u}^{n+1}
+ * - \bm{u}_b) \f$. The class computes the coefficient \f$
+ * \frac{\chi}{\kappa}$ of the fluid velocity \f$ \bm{u}^{n+1} \f$ for the
+ * variable-coefficient INS solvers of INSVCStaggeredHierarchyIntegrator. This
+ * is done in the BrinkmanPenalizationRigidBodyDynamics::demarcateBrinkmanZone
+ * method. Here \f$ \chi \f$ is the body indicator function and \f$\kappa \sim
+ * \Delta t/ \rho \ll 1 \f$ is the vanishing permeability of the body. The
+ * rigid body velocity \f$\bm{u}_b\f$ is computed through Newton's law of
+ * motion by netting the hydrodynamic and external forces and torques on the
+ * body in the BrinkmanPenalizationRigidBodyDynamics::computeBrinkmanVelocity
+ * method. A simple forward-Euler scheme is imployed for the second-law of
+ * motion.
+ *
+ * For further information on applications of this class see
+ * https://arxiv.org/abs/1904.04078.
+ */
+class BrinkmanPenalizationRigidBodyDynamics : public BrinkmanPenalizationStrategy
+{
+    ////////////////////////////// PUBLIC ////////////////////////////////////////
+
+public:
+    /*
+     * \brief Constructor of the class.
+     */
+    BrinkmanPenalizationRigidBodyDynamics(std::string object_name,
+                                          SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > ls_solid_var,
+                                          SAMRAI::tbox::Pointer<IBAMR::AdvDiffHierarchyIntegrator> adv_diff_solver,
```

Comment:
This is the *level set* solver, right? Should we name it as such?

Reply:
This is used as part of IBLevelSetHI, but I think it is more appropriate to name it by the numerical technique.

Reply:
Actually it is used by INSVCHI and not by BLevelSetHI directly.

Reply:
I am ambivalent about the name so I kept it as-is since it works with rigid bodies.